diff --git a/node_modules/js-green-licenses/build/src/checker.js b/node_modules/js-green-licenses/build/src/checker.js
index 79a88a3..d38efa4 100644
--- a/node_modules/js-green-licenses/build/src/checker.js
+++ b/node_modules/js-green-licenses/build/src/checker.js
@@ -68,6 +68,7 @@ const DEFAULT_GREEN_LICENSES = [
     'Apache-2.0',
     'Artistic-1.0',
     'Artistic-2.0',
+    'BlueOak-1.0.0',
     'BSD-2-Clause',
     'BSD-3-Clause',
     'BSL-1.0',
@@ -228,6 +229,11 @@ class LicenseChecker extends events_1.EventEmitter {
         const spec = `${packageName}@${versionSpec}`;
         if (this.failedPackages.has(spec))
             return;
+        if (versionSpec.startsWith('npm:')) {
+            packageName = versionSpec.split('@')[0]
+            packageName = packageName.replace('npm:', '')
+            versionSpec = versionSpec.split('@')[1]
+        }
         // remove tilde/caret to check for an exact version, ^0.5.0-rc.0 becomes 0.5.0-rc.0
         const version = versionSpec.replace(/^[^~]/, '');
         // if the dependency is a local package then skip verification at this step. will be checked independently
